Improve update channel layout to be a popup menu

MorrisJobke opened the pull request:
* now the entries are in a popup menu
* if a non-default channel is used the entry will be at the bottom
* channel descriptions are embeded into the menu
* channels are now localized
* fixes #13806
* [x] needs https://github.com/nextcloud/nextcloud-vue/pull/249
* [x] rebase and squash fixups once #14177 is merged

This is how it looks (normal menu, after and update, with a non-default entry).

![bildschirmfoto 2019-02-13 um 15 23 51](https://user-images.githubusercontent.com/245432/52718310-6d36e680-2fa3-11e9-855a-6a87034ccbb8.png)

![bildschirmfoto 2019-02-12 um 18 09 20](https://user-images.githubusercontent.com/245432/52654225-de1dc600-2ef1-11e9-9025-558cef2b70ea.png)

![bildschirmfoto 2019-02-13 um 14 35 25](https://user-images.githubusercontent.com/245432/52715494-dcf5a300-2f9c-11e9-85fd-a6a7cbd366af.png)

MorrisJobke committed:
Improve update channel layout to be a popup menu

* now the entries are in a popup menu
* if a non-default channel is used the entry will be at the bottom
* channel descriptions are embeded into the menu
* use nextcloud-vue componenets for the channel selector
* update nextcloud-vue dependency from 0.6.1 to 0.6.3
* fixes #13806

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

ChristophWurst commented:
>     * use better icons @nextcloud/designers ideas welcome

* :elephant: production
* :horse: stable
* :tiger:  beta
* :clown_face: git

MorrisJobke commented on settings/css/settings.scss:
```diff
@@ -962,6 +962,10 @@ span.version {
 
+#version.section {
+	border-bottom: none;
+}
```
This one fixes the border that was shown in the update section, because in the version part it is not the last element (because the update notification injects another element below.

MorrisJobke replied:
Moved to #14177

jancborchardt commented:
The icons for beta and git are already very fitting! :)

- [ ] For production, I’d use `icon-star` and for stable `icon-checkmark`.
- [ ] The word of the update channel (like "Production") should also be clickable so it’s a bigger area. So ideally the whole clickable block is `✔ Production ▾`.

MorrisJobke commented:
> The icons for beta and git are already very fitting! :)
> 
> * [x]  For production, I’d use `icon-star` and for stable `icon-checkmark`.
> * [x]  The word of the update channel (like "Production") should also be clickable so it’s a bigger area. So ideally the whole clickable block is `✔ Production ▾`.

Done. Looks now like this:

![bildschirmfoto 2019-02-12 um 23 40 50](https://user-images.githubusercontent.com/245432/52673127-1805c100-2f20-11e9-8fa6-d073051d679f.png)


Ready for review - I will squash once reviews are in.

skjnldsv commented on apps/updatenotification/src/components/root.vue:
```diff
@@ -63,22 +63,59 @@
+					{{ localizedChannelName }}
+					<span class="icon-triangle-s"></span>
+				</span>
+				<div class="update-channel-menu popovermenu bubble menu menu-center" v-bind:class="{ 'show-menu': openedUpdateChannelMenu}"><ul>
```
Please use the popovermenu component :)
https://github.com/nextcloud/nextcloud-vue

MorrisJobke replied:
Is there a bit more documentation about the needed properties and names than `PopoverMenu` as listed component? 🙈

skjnldsv commented on apps/updatenotification/src/components/root.vue:
```diff
@@ -63,22 +63,59 @@
+		<h3 class="update-channel-selector">
+			{{ t('updatenotification', 'Update channel:') }}
+			<div class="update-menu">
+				<span class="icon-update-menu" @click="toggleUpdateChannelMenu" v-bind:class="{ 'icon-star': currentChannel === 'production', 'icon-checkmark': currentChannel === 'stable', 'icon-category-customization': currentChannel === 'beta', 'icon-rename': isNonDefaultChannel }">
```
I would remove the icon from the selected option, just `Update-Channel: Production ▾`

MorrisJobke replied:
Makes sense IMO - @jancborchardt proposed this before.

nickvergessen commented:
I would say "stable" should be at the top, as that is the recommended option to choose for most cases.

MorrisJobke committed:
Fix corners that are cut in popup menu in Safari

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

MorrisJobke commented:
Okay - this is now ready for a final review. This includes two other PRs and I will rebase once both are included. Until then the screenshots and checked in JS assets are for testing.

I also updated the screenshots of the OP.

skjnldsv commented on apps/updatenotification/src/components/root.vue:
```diff
@@ -63,22 +63,24 @@
+					{{ localizedChannelName }}
+					<span class="icon-triangle-s"></span>
+				</span>
+				<div class="popovermenu bubble menu menu-center" v-bind:class="{ 'show-menu': openedUpdateChannelMenu}">
```
you do not need the bubble class since ages I think :)

MorrisJobke replied:
Done.

skjnldsv commented on apps/updatenotification/src/components/root.vue:
```diff
@@ -241,6 +231,66 @@
+					longtext: t('updatenotification', 'The most recent stable version. It is suited for regular use and will always update to the latest major version.'),
+					icon: 'icon-checkmark',
+					active: this.currentChannel === 'stable',
+					action: function() { self.changeReleaseChannel("stable") }
```
You can:
1. use arrow functions and drop the self
    action: () => { this.changeReleaseChannel("stable") }
2. use dedicated functions inline
    `action: this.changeReleaseChannelToStable`

skjnldsv replied:
ps: I love the second, it's cleaner to read and the code is compiled anyway. ;)

MorrisJobke replied:
1. ah - didn't knew that 
2. I would then need to define all of them - in this way it's a bit more flexible and clear that is does the same IMO

skjnldsv replied:
Yes, but this is 3 functions that can be easily understood :)
It's better to have complicated data in another function than inside an object that is already pretty full 
;)

MorrisJobke replied:
Okay - adjusted.

skjnldsv commented on apps/updatenotification/src/components/root.vue:
```diff
@@ -241,6 +231,66 @@
+					longtext: t('updatenotification', 'The most recent stable version. It is suited for regular use and will always update to the latest major version.'),
+					icon: 'icon-checkmark',
+					active: this.currentChannel === 'stable',
+					action: function() { self.changeReleaseChannel("stable") }
```
Also single quotes pls :)

skjnldsv commented on apps/updatenotification/src/components/root.vue:
```diff
@@ -63,22 +63,24 @@
+		<h3 class="update-channel-selector">
+			{{ t('updatenotification', 'Update channel:') }}
+			<div class="update-menu">
+				<span class="icon-update-menu" @click="toggleUpdateChannelMenu">
```
If you want you can also use the clickoutside plugin to close on..well.. click outside of the popover :)

MorrisJobke replied:
I would leave it as it is for now.

skjnldsv reviewed (dismissed):
Few nitpicks/comments :)

MorrisJobke commented:
I fixed the comments by @skjnldsv and this is now fully rebased and squashed.

I also adjusted the text for production to be this:

> Will always provide the latest patch level, but not update to the next major release immediately. That update usually happens with the second minor release (x.0.2) and only if the instance is already on the latest minor version.

Does this make sense? It should reflect the planned changes from #8799 that is then done to the updater_server.

skjnldsv reviewed (approved):
Awesome work!! :)
